File: praca-inzynierska-frontend/src/app/app/model/dto/StatisticsYearDto.ts
import { EmployeeDto } from './EmployeeDto';
import { PositionDto } from './PositionDto';
import { StatisticsEmployeeYearDto } from './StatisticsEmployeeYearDto';
import { StatisticsPositionYearDto } from './StatisticsPositionYearDto';

export class StatisticsYearDto {

    constructor(
        public year: number,
        public positionStatistics: StatisticsPositionYearDto[],
        public employeeStatistics: StatisticsEmployeeYearDto[]
    ) {}

    sort() {
        this.positionStatistics.sort((a, b) => PositionDto.compare(a.position, b.position))
        this.employeeStatistics.sort((a, b) => EmployeeDto.compare(a.employee, b.employee))
        this.employeeStatistics.forEach(e => e.sort())
    }

    static copyOf(other: StatisticsYearDto): StatisticsYearDto {
        return new StatisticsYearDto(other.year, 
            other.positionStatistics.map(p => StatisticsPositionYearDto.copyOf(p)),
            other.employeeStatistics.map(e => StatisticsEmployeeYearDto.copyOf(e))
        )
    }
}

File: praca-inzynierska-backend/application/src/main/kotlin/jwer/pracainzynierskabackend/model/dto/StatisticsYearDto.kt
package jwer.pracainzynierskabackend.model.dto

import jwer.pracainzynierskabackend.model.entity.Shift

data class StatisticsYearDto(
        val year: Int,
        val positionStatistics: List<StatisticsPositionYearDto>,
        val employeeStatistics: List<StatisticsEmployeeYearDto>
) {

    fun add(shift: Shift) {
        val month = shift.period.start.monthValue - 1
        positionStatistics.find { it.position.id == shift.position.id }!!.totalMonths[month] += shift
        employeeStatistics.find { it.employee.employeeId == shift.employee.id }!!.total += shift
    }

}